File: codes/python/factorial.py
from numbers import Number

print("\n# Factorial(n) in Python")


def factorial_v1(any_number):
    if (isinstance(any_number, Number) == True):
        raise Exception("Argument should be a number")
    if (any_number < 0):
        raise Exception("Argument should be >= 0")
    if (any_number == 0):
        return 1
    result = 1
    i = any_number
    while (True):
        result *= i
        if (i <= 1):
            break
        i -= 1
    return result


def factorial_v2(any_number):
    if (isinstance(any_number, Number) == True):
        raise Exception("Argument should be a number")
    if (any_number < 0):
        raise Exception("Argument should be >= 0")
    if (any_number == 0):
        return 1
    result = 1
    i = any_number
    while (i >= 1):
        result *= i
        i -= 1
    return result


def factorial_v3(any_number):
    if (isinstance(any_number, Number) == True):
        raise Exception("Argument should be a number")
    if (any_number < 0):
        raise Exception("Argument should be >= 0")
    if (any_number == 0):
        return 1
    result = 1
    for i in range(any_number, 0, -1):  # (start, stop, step)
        result *= i
    return result


def factorial_v4(any_number):
    if (isinstance(any_number, Number) == True):
        raise Exception("Argument should be a number")
    if (any_number < 0):
        raise Exception("Argument should be >= 0")
    if (any_number == 0):
        return 1
    return (any_number * factorial_v4(any_number - 1))


print('# using factorial function "factorial_v1"')
print(f"Factorial(5): {factorial_v1(5)}")
# Factorial(5): 120

print('# using factorial function "factorial_v2"')
print(f"Factorial(5): {factorial_v2(5)}")
# Factorial(5): 120

print('# using factorial function "factorial_v3"')
print(f"Factorial(5): {factorial_v3(5)}")
# Factorial(5): 120

print('# using factorial function "factorial_v4"')
print(f"Factorial(5): {factorial_v4(5)}")
# Factorial(5): 120

File: codes/javascript/src/factorial.js
console.log("\n// Factorial(n) in JavaScript");

const factorialV1 = (anyNumber) => {
    if (typeof anyNumber !== "number") throw new Error("Argument should be a number");
    if (anyNumber < 0) throw new Error("Argument should be >= 0");
    if (anyNumber === 0) return 1;
    let result = 1;
    let i = anyNumber;
    while (true) {
        result *= i;
        if (i <= 1) break;
        i -= 1;
    }
    return result;
};

const factorialV2 = (anyNumber) => {
    if (typeof anyNumber !== "number") throw new Error("Argument should be a number");
    if (anyNumber < 0) throw new Error("Argument should be >= 0");
    if (anyNumber === 0) return 1;
    let result = 1;
    let i = anyNumber;
    while (i >= 1) {
        result *= i;
        i -= 1;
    }
    return result;
};

const factorialV3 = (anyNumber) => {
    if (typeof anyNumber !== "number") throw new Error("Argument should be a number");
    if (anyNumber < 0) throw new Error("Argument should be >= 0");
    if (anyNumber === 0) return 1;
    let result = 1;
    for (let i = anyNumber; i >= 1; i -= 1) {
        result *= i;
    }
    return result;
};

const factorialV4 = (anyNumber) => {
    if (typeof anyNumber !== "number") throw new Error("Argument should be a number");
    if (anyNumber < 0) throw new Error("Argument should be >= 0");
    if (anyNumber === 0) return 1;
    return (anyNumber * factorialV4(anyNumber - 1));
};

console.log(`// using factorial function "factorialV1"`);
console.log(`Factorial(5): ${factorialV1(5)}`);
// Factorial(5): 120

console.log(`// using factorial function "factorialV2"`);
console.log(`Factorial(5): ${factorialV2(5)}`);
// Factorial(5): 120

console.log(`// using factorial function "factorialV3"`);
console.log(`Factorial(5): ${factorialV3(5)}`);
// Factorial(5): 120

console.log(`// using factorial function "factorialV4"`);
console.log(`Factorial(5): ${factorialV4(5)}`);
// Factorial(5): 120
